Headers/state_chm_mod.F90

```diff
@@ -2395,21 +2395,25 @@ SUBROUTINE Init_Hg_Simulation_Fields( Input_Opt, State_Chm, State_Grid,    &
 
     ! Index array: Hg0 species # <--> Hg0 category #
     ALLOCATE( State_Chm%Hg0_Id_List( State_Chm%N_Hg_CATS ), STAT=RC )
+    CALL GC_CheckVar( 'State_Chm%Hg0_Id_List', 0, RC )
     IF ( RC /= GC_SUCCESS ) RETURN
     State_Chm%Hg0_Id_List = 0
 
     ! Index array: Hg2 species # <--> Hg0 category #
     ALLOCATE( State_Chm%Hg2_Id_List( State_Chm%N_Hg_CATS ), STAT=RC )
+    CALL GC_CheckVar( 'State_Chm%Hg2_Id_List', 0, RC )
     IF ( RC /= GC_SUCCESS ) RETURN
     State_Chm%Hg2_Id_List = 0
```

Comment:
@yantosca if I am understanding the behavior correctly, adding in GC_Checkvar will throw an error if the variable isn't successfully allocated. So would the following if statement ever get tripped? Since failure mode would now halt execution?

Reply:
So `GC_CheckVar` just defines an error message depending on whether or not the error happened at allocation, registration, or deallocation.  Then it calls `GC_Error`, which prints the error message as well as forcing `RC = GC_FAILURE`.  Then RC comes back from `GC_CheckVar` as failure, which will keep returning the program control to the top level where it'll stop gracefully.

```F90
!EOC
!------------------------------------------------------------------------------
!                  GEOS-Chem Global Chemical Transport Model                  !
!------------------------------------------------------------------------------
!BOP
!
! !IROUTINE: GC_CheckVar
!
! !DESCRIPTION: Wrapper routine for GC\_Error.  Prints an error message
!  if there is an allocation or registration error.  This is intended to
!  be called from the state initialization method (e.g. Init\_State\_Met).
!\\
!\\
! !INTERFACE:
!
  SUBROUTINE GC_CheckVar( Variable, Operation, RC )
!
! !INPUT PARAMETERS:
!
    CHARACTER(LEN=*), INTENT(IN)    :: Variable   ! Name of variable to check
    INTEGER,          INTENT(IN)    :: Operation  ! 0=Allocate
                                                  ! 1=Register
                                                  ! 2=Deallocate
!
! !OUTPUT PARAMETERS:
!
    INTEGER,          INTENT(INOUT) :: RC         ! Success or failure
!
! !REMARKS:
!  You also need to add an
!    IF ( RC /= GC_SUCCESS ) RETURN
!  from the calling routine for proper error handling.
!
! !REVISION HISTORY:
!  27 Jun 2017 - R. Yantosca - Initial version
!  See https://github.com/geoschem/geos-chem for complete history
!EOP
!------------------------------------------------------------------------------
!BOC
!
! !LOCAL VARIABLES:
!
  ! Strings
  CHARACTER(LEN=255) :: ErrMsg, ThisLoc

  !=========================================================================
  ! Initialize
  !=========================================================================

  ! Define error message
  SELECT CASE( Operation )
     CASE( 1 )
        ErrMsg = 'Could not register '   // TRIM( Variable ) // '!'
     CASE( 2 )
        ErrMsg = 'Could not deallocate ' // TRIM( Variable ) // '!'
     CASE DEFAULT
        ErrMsg = 'Could not allocate '   // TRIM( Variable ) // '!'
  END SELECT

  ! Define location string
  ThisLoc   = ' -> at GC_CheckVar (in Headers/errcode_mod.F90)'

  !=========================================================================
  ! Display error message if necessary
  !=========================================================================
  IF ( RC /= GC_SUCCESS ) THEN
     CALL GC_Error( ErrMsg, RC, ThisLoc )
  ENDIF

  END SUBROUTINE GC_CheckVar
```

Reply:
Got it, thanks!